Identity Agent Guide (#1149)
TBD54566975/developer.tbd.website: site/docs/web5/build/decentralized-web-nodes/using-identity-agents.mdx

Reviewer: it's still not clear how to i reuse my agent outside of this program

Reviewer 2: The program would be an identity management app so you would interact with the agent through some front end app that is party of the identity management app.

Reviewer: I think we should clearly specify that this is a separate application and not in the same program where you call `Web5.connect()`

Reviewer: I see you state it later on. Up to you if you think we should call it out early in the guide as well

Author: okay, just added some more clarity the line in the beginning to show that they are separate apps: `This guide will walk you through creating an Identity Agent as a standalone application to manage multiple user identities and connect to a separate Web5 application.`

---
sidebar_position: 9
---

import CodeSnippet from '@site/src/components/CodeSnippet';

# Using Identity Agents

Identity Agents are a specialized type of [agent](https://developer.tbd.website/docs/web5/learn/agents/) that act as personal identity managers.  Similar to how a password manager securely stores login credentials for different web apps, Identity Agents manage your [DIDs](https://developer.tbd.website/docs/web5/learn/decentralized-identifiers). 

> ## Doesn't Web5.connect() create a DID for me?

Reviewer: nit: consider making this a regular sentence vs a header. it seems a bit out of place looking at it with the other headers

 
Yes, the `Web5.connect()` function generates a **new** identity and a local User Agent. However, Identity Agents allow you to connect a web app to an **existing** identity. This is useful for maintaining a cohesive online presence. 

This guide will walk you through creating an Identity Agent to manage multiple user identities and connect to a Web5 application.

<details>
<summary>Prerequisites</summary>

**Install the following packages**

```bash
npm i @web5/dids @web5/identity-agent @web5/api
```

**Import the following modules**
    
```js
import { DidIonMethod } from '@web5/dids';
import { IdentityAgent } from '@web5/identity-agent';
import { Web5, getTechPreviewDwnEndpoints } from '@web5/api';
```
    
</details>

## Initialize the Agent
Start by creating an instance of the `IdentityAgent`.

<CodeSnippet functionName="createIdentityAgent" />

## Prompt Users to Authenticate

> The first time you launch an Identity Agent, you will be prompted to authenticate with a one-time passphrase for security purposes.

Reviewer: this is confusing since this doesn't happen with any of the code we're showing. Should we tell them they should build that into their identity agent app code?

<CodeSnippet functionName="authenticateIdentityAgent" />

> ## Connect User DIDs to DWNs

Reviewer: ```suggestion
## Connect DIDs to DWNs
```

Next, enable the Identity Agent to manage user DIDs by connecting each DID to a [DWN](https://developer.tbd.website/docs/web5/learn/decentralized-web-nodes/). Below is an example using TBD-hosted DWN endpoints:

<CodeSnippet functionName="getDwnEndpoints" />

<details>
<summary>Expected Output of didOptions</summary>

```js
{
  keySet: { verificationMethodKeys: [ [Object], [Object] ] },
  services: [
    {
      id: '#dwn',
      serviceEndpoint: [Object],
      type: 'DecentralizedWebNode'
    }
  ]
}
```
    
</details>

## Create Identities
Now that `didOptions` contains cryptographic key pairs and service endpoints, you can create as many identities as necessary. 

Each identity, linked to a unique DID, represents and compartmentalizes different aspects of a user. For example, a user can have an identity for social media interactions and another for professional engagements.

<CodeSnippet functionName="createSocialMediaAndCareerIdentity" />

:::note
* The `name` field serves as a friendly name for reference. 

> * The `kms` field stands for Key Management Service and can securely store and manage the cryptographic keys associated with a DID. If you don't specify a particular key management service, such as AWS Key Management Service or Google Cloud Key Management, Web5 will default to an in-memory key manager.

Reviewer: ```suggestion
* The `kms` field stands for [Key Management Service](/docs/web5/build/decentralized-identifiers/key-management) and can securely store and manage the cryptographic keys associated with a DID. If you don't specify a particular key management service, such as AWS Key Management Service, Web5 will default to an in-memory key manager.
```

I dont think Google's is supported

While the `Web5.connect()` function creates a local User Agent and a **new** identity, Identity Agents allow you to connect a web app to an **existing** identity.

This guide provides an overview of how to enable an Identity Agent to manage multiple user identities on behalf of your application.

<details>
<summary>Prerequisites</summary>

**Install the following packages**

```bash
npm i @web5/dids @web5/identity-agent @web5/api
```

**Import the following modules**
    
```js
import { DidIonMethod } from '@web5/dids';
import { IdentityAgent } from '@web5/identity-agent';
import { Web5, getTechPreviewDwnEndpoints } from '@web5/api';
```
    
</details>

## Initialize the Agent
Start by creating an instance of the `IdentityAgent`.

<CodeSnippet functionName="createIdentityAgent" />

## Prompt Users to Authenticate

> The first time an end user launches an application that uses an Identity Agent, they'll need to authenticate with a one-time passphrase for security purposes. 

Reviewer 2: The front-end of the identity agent app.

In other words, the app IS the identity agent.   It isn't "connecting" to anything -- the point of an identity app is to manage identities and that's what the `IdentityAgent` implementation was built for.  An example would be the mobile Identity App that Tim is building but you could also have a desktop electron identity app.  Since this is an example, it would be a browser based version.

> The first time an end user launches an application that uses an Identity Agent, they'll need to authenticate with a one-time passphrase for security purposes. 

Reviewer: so the identity agent creation and the web5 connection shouldn't be shown in the same program. these are 2 different apps.

i think we're missing the glue piece. if I'm a web5 app developer, what do i do to engage with an identity app such that I can then pass it into `connect()`?

> The first time an end user launches an application that uses an Identity Agent, they'll need to authenticate with a one-time passphrase for security purposes. 

Author: Thanks for the clarification, Frank!

+1 to Angie's question. For me, I understand that the Identity Agent's purpose is to manage identities. 

But after I wrote the initial draft I did start (over)thinking because I want to provide clarity to our community..and I think I went off course. The question for me is: A Web5 app needs to obtain identity information from the identity agent to interact with the user's decentralized identity. How does it do that?

Here's my understanding of the workflow based on this conversation: 
1. The user opens the identity agent app (which could be a mobile app, a desktop application, or a browser-based application).
2. The **identity agent** prompts the user to authenticate, by entering a passphrase or biometric auth or whatever. (It says at first launch. Does this mean the first time the user ever uses the identity agent?)
3.  Once authenticated, the user can view, manage, or create new identities (social, professional, or whatever)
4. User finds a Web5 app that they want to use
5. The Web5 app says..give me your DID so you can start using it
6. User says okay I want to use my social Identity so I'm going to paste in the DID associated with my social identity into this Web5 app. (I think maybe what I was imagining was their would be some code or an automated process within our identity agent, so that we could integrate the Web5 app with the identity agent and you won't have copy and paste your did into the Web5 app..and maybe this is the glue that Angie is looking for)
7. Then the Web5 app, takes your the DID that your provided and connects to the Web5 ecosystem.

Thoughts? @frankhinek

> The first time an end user launches an application that uses an Identity Agent, they'll need to authenticate with a one-time passphrase for security purposes. 

Reviewer: This is from August, not sure if it's still 100% accurate, but may help with some details
https://hackmd.io/@frankhinek/ryge2ayh3

> The first time an end user launches an application that uses an Identity Agent, they'll need to authenticate with a one-time passphrase for security purposes. 

Author: Thank you! Yes, that helped with my question about first launch.

But I still want to be able to answer this question you had " if I'm a web5 app developer, what do i do to engage with an identity app such that I can then pass it into connect()". 

Because for right now, all I know is that the end user can take the DID from the identity agent and copy/paste it into the Web5 app.

> The first time an end user launches an application that uses an Identity Agent, they'll need to authenticate with a one-time passphrase for security purposes. 

Reviewer 2: Hey @blackgirlbytes -- sent a number of files over to see if any address the question adequately.

Once we figure out if any of that info helps, we can post / revise here 😄

> The first time an end user launches an application that uses an Identity Agent, they'll need to authenticate with a one-time passphrase for security purposes. 

Author: Thank you! Read the items and talked to frank on the phone. 

Web5 app has a QR code section that says connect to Identity agent — you take your identity agent and you scan that QR code and then the identity agent adds that profile/identity to the app.

Or could be a deep link and you might type in something to add your identity.

Going to take my final pass at the documentation. 

![image](https://github.com/TBD54566975/developer.tbd.website/assets/22990146/442d75d8-dd31-46b1-bb09-0c6000825e05)

---
sidebar_position: 9
---

import CodeSnippet from '@site/src/components/CodeSnippet';

# Using Identity Agents

[Agents](https://developer.tbd.website/docs/web5/learn/agents/) manage [DIDs](https://developer.tbd.website/docs/web5/learn/decentralized-identifiers), data, and communication between [DWNs](https://developer.tbd.website/docs/web5/learn/decentralized-web-nodes). Identity Agents are a type of agent that specialize in managing multiple decentralized identities, making it easier to use existing identities across various Web5 applications.

> 

Author: added these lines based on Frank's feedback

> While the `Web5.connect()` function creates a local User Agent and a **new** identity, Identity Agents allow you to connect a web app to an **existing** identity.

Author: added these lines based on Frank's feedback

This guide provides an overview of how to use an identity agent to create and manage multiple identities.

<details>
<summary>Prerequisites</summary>

**Install the following packages**

```bash
npm i @web5/dids @web5/identity-agent @web5/api
```

**Import the following modules**
    
```js
import { DidIonMethod } from '@web5/dids';
import { IdentityAgent } from '@web5/identity-agent';
import { Web5, getTechPreviewDwnEndpoints } from '@web5/api';
```
    
</details>

## Initialize the Agent
To begin, create an instance of `IdentityAgent`.

<CodeSnippet functionName="createIdentityAgent" />

## Authenticate with a passphrase

> For security purposes, the identity agent prompts the end user to enter a secure passphrase when it initially launches.

Reviewer: does this mean, the developer's app should prompt the user for a passphrase?

<CodeSnippet functionName="authenticateIdentityAgent" />

## Configure your DID
If you'd like to associate DWNs with a DID, you can do so. In the following example, we'll use the TBD-hosted DWN endpoints:

<CodeSnippet functionName="getDwnEndpoints" />

<CodeSnippet functionName="createDidOptions" />

<details>

> <summary>Expected Output</summary>

Reviewer: output of what? maybe rename this to make it clear this is what's stored in `didOptions`

    
```js
{
  keySet: { verificationMethodKeys: [ [Object], [Object] ] },
  services: [
    {
      id: '#dwn',
      serviceEndpoint: [Object],
      type: 'DecentralizedWebNode'
    }
  ]
}
```
    
</details>

## Create Identities

> Now that `didOptions` contains cryptographic key pairs and service endpoints, you can create as many identities as necessary. Each identity, linked to a unique DID, represents and compartmentalizes different aspects of a user. For example, you can have a identity for social media interactions and another for professional engagements.

Reviewer: cool!

---
sidebar_position: 9
---

import CodeSnippet from '@site/src/components/CodeSnippet';

# Using Identity Agents

Identity Agents are a specialized type of [agent](https://developer.tbd.website/docs/web5/learn/agents/) that act as personal identity managers.  Similar to how a password manager securely stores login credentials for different web apps, Identity Agents manage your [DIDs](https://developer.tbd.website/docs/web5/learn/decentralized-identifiers). 
 

> While the `Web5.connect()` function generates a **new** identity and a local User Agent, Identity Agents allow you to connect a web app to an **existing** identity. This is useful for maintaining a cohesive online presence. 

Reviewer: ```suggestion
By default, the `Web5.connect()` function generates a **new** identity; however, Identity Agents allow you to connect to a Web5 app with an **existing** identity. This is useful for maintaining a cohesive online presence. 
```

<CodeSnippet functionName="createSocialMediaAndCareerIdentity" />

:::note
* The `name` field serves as a friendly name for reference. 
* The `kms` field stands for [Key Management Service](/docs/web5/build/decentralized-identifiers/key-management) and can securely store and manage the cryptographic keys associated with a DID. If you don't specify a particular key management service, such as AWS Key Management Service, Web5 will default to an in-memory key manager.
:::

## Use your Identity in a Web5 Application
Once your Identity Agent is set up to manage multiple identities, you can use it to help you login to a Web5 application. Here's the typical workflow: 

1. The Web5 application may display a QR code or a deep link for login purposes.
2. Use the Identity Agent to scan the QR code or click the deep link to establish a connection between your Identity Agent and the Web5 application.
3. The Identity Agent then prompts you to select the identity you prefer to use for authentication. 
 
## Connect to Web5
Finally, the Web5 application can use the DID linked to your identity, so that you can interact within the Web5 application as your chosen identity. 

> ```js

Reviewer: can you add this as a code snippet vs hard-coding? Can skip adding tests if this poses a problem with our current setup

Author: done! :)

---
sidebar_position: 9
---
# Using Identity Agents

> Identity [agents](http://localhost:3000/docs/web5/learn/agents) manage [DIDs](https://developer.tbd.website/docs/web5/learn/decentralized-identifiers), data, and communication between [DWNs](https://developer.tbd.website/docs/web5/learn/decentralized-web-nodes). 

Reviewer 2: ```suggestion
Identity [agents](https://developer.tbd.website/docs/web5/learn/agents/) manage [DIDs](https://developer.tbd.website/docs/web5/learn/decentralized-identifiers), data, and communication between [DWNs](https://developer.tbd.website/docs/web5/learn/decentralized-web-nodes). 
```

Author: oop thank you!

> For convenience, the `Web5.connect()` function uses a default identity agent. However, Web5 also provides the ability for users to utilize their own custom agent. 

Reviewer 2: I would also suggest highlighting something like:

Identity Agents are a specialized kind of agent that are primarily designed to enable people to manage multiple  decentralized identities and to make it easy to use your existing identities across multiple web apps.

When calling `Web5.connect()`, rather than creating a local User Agent and a new decentralized identity, with an Identity Agent you can connect the web app to an existing identity.

Author: okay, will add!

This guide provides an overview of how to create an identity agent, import a DID into the agent, and connect the agent to a Web5 app.

<details>
<summary>Prerequisites</summary>

**Install the following packages**

```bash
npm i @web5/dids @web5/user-agent @web5/api
```

**Import the following modules**
```js
import { DidIonMethod, DidKeyMethod } from '@web5/dids';
import { Web5UserAgent } from '@web5/user-agent';
import { Web5, getTechPreviewDwnEndpoints } from '@web5/api';
```

> This guide focuses on utilizing `DidIonMethod` and `DidKeyMethod` for setting up an identity agent. Currently, DID methods like the `DidDhtMethod` are not supported in this context.

Reviewer: ```suggestion
This guide focuses on utilizing `DidIonMethod` and `DidKeyMethod` for setting up an identity agent. Currently,  `DidDhtMethod` is not supported in this context.
```

</details>

## Initialize the Agent
Create an instance of `Web5UserAgent` to begin setting up the Identity Agent
```javascript
const agent = await Web5UserAgent.create();
```

## Authenticate with a passphrase

> For security purposes, the identity agent prompts the end user to enter a secure passphrase when it initially launches. In this example, `insecure-static-phrase` is used as a placeholder.

Reviewer: ```suggestion
For security purposes, the identity agent prompts the end user to enter a secure passphrase when it initially launches. For now, `insecure-static-phrase` must be used.
```
